src/contrib/mlir/core/pass/ngraph_dialect.cpp

```diff
@@ -0,0 +1,108 @@
+using llvm::ArrayRef;
+
+using namespace ngraph;
+using namespace ngraph::runtime::ngmlir;
+
+ngraph::pass::NgDialectConversionPass::NgDialectConversionPass(const ngraph::op::CompiledKernel* compiledKernel)
+    :m_compiledKernel(compiledKernel)
+{
+
+}
+
+void ngraph::pass::NgDialectConversionPass::buildNgDialect()
+{
+    const NodeVector& subGraph = m_compiledKernel->get_node_list();
+
+    for (auto np : subGraph)
+    {
+        auto it = opDispatcher.find(TI(*np));
+        if (it == opDispatcher.end())
+        {
+            throw unsupported_op{std::string{"The MLIR backend doesn't currently implement the '"} +
+                                 np->description() + "' operation"};
+        }
+        mlir::Operation* op = it->second(*this, np.get());
+        // This assumes simple 1:1 mapping between output edges and generated MLIR op results
+        // If the mapping is more complex, the create_op helper can return null operation
+        // and handles populating the value map itself
+        if (op)
+        {
+            for (auto i = 0; i < op->getNumResults(); i++)
+            {
+                mlir::Value* result = op->getResult(i);
+                if (result)
+                {
+                    updateTensorValue(np->get_output_tensor_ptr(i).get(), result);
+                }
+            }
+        }
+    }
+    createReturn();
+}
+
+void ngraph::pass::NgDialectConversionPass::runOnModule()
+{
+    TypeList argsTypeList, resultTypeList;
+
+    // Retrieve input and output tensors.
+    const auto& kernelInputs = m_compiledKernel->get_arguments();
+    const auto& kernelOutput = m_compiledKernel->get_kernel_outputs();
+    NGRAPH_CHECK(kernelInputs.size() != 0, "Cannot have empty inputs list");
+    NGRAPH_CHECK(kernelOutput.size() != 0, "Cannot have empty outputs list");
+
+    for (auto input : kernelInputs)
+    {
+        argsTypeList.push_back(getMlirType(input.get()));
+    }
+
+    for (auto output : kernelOutput)
+    {
+        resultTypeList.push_back(getMlirType(output.get()));
+    }
+
+    auto funcType = mlir::FunctionType::get(argsTypeList, resultTypeList, &m_context);
+    auto function = mlir::FuncOp::create(mlir::UnknownLoc::get(&m_context), "main", funcType);
+    function.addEntryBlock();
+
+    // populate Tensor->Value maps
+    int i = 0;
+    for (auto input : kernelInputs)
+    {
+        mlir::Value* arg = function.getArgument(i);
+        TensorInfo tensorInfo{arg};
+        m_tensorToValueMap.insert(TensorToInfo(input->get_output_tensor_ptr().get(), tensorInfo));
+        i++;
+    }
+
+    // create builder
+    m_builder = std::unique_ptr<mlir::OpBuilder>(new mlir::OpBuilder(function.getBody()));
+    ngraph::pass::NgDialectConversionPass::buildNgDialect();
+}
+
+std::unique_ptr<mlir::Pass> ngraph::pass::CreateNgDialectConversionPass(const ngraph::op::CompiledKernel* compiledKernel)
+{
+	return std::make_unique<ngraph::pass::NgDialectConversionPass>(compiledKernel);
+}
```

Comment:
We would need to register the pass for debugging and to be able to use it in `ngraph-opt` with a flag. Example:
https://github.com/tensorflow/mlir/blob/master/lib/Transforms/LoopUnroll.cpp#L37-L39
https://github.com/tensorflow/mlir/blob/master/lib/Transforms/LoopUnroll.cpp#L191

Reply:
Just a side comment: this pass won't run through ngraph-opt though. We could add support to de-serialize a graph though and run this pass.